sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py

```diff
@@ -155,3 +197,158 @@ def _ns_to_iso8601_string(nanoseconds: int) -> str:
     dt_microseconds = timedelta(microseconds=microseconds)
     dt_with_microseconds = dt + dt_microseconds
     return dt_with_microseconds.isoformat()
+
+
+# Filtering
+
+def _update_filter_configuration(etag: str, config_bytes: bytes):
+    # Clear projection map
+    _clear_quickpulse_projection_map()
+    seen_ids = set()
+    # config is a byte string that when decoded is a json
+    config = json.loads(config_bytes.decode("utf-8"))
+    metric_infos: Dict[TelemetryType, List[DerivedMetricInfo]] = {}
+    for metric_info_dict in config.get("Metrics", []):
+        metric_info = DerivedMetricInfo.from_dict(metric_info_dict)
+        # Skip duplicate ids
+        if metric_info.id in seen_ids:
+            continue
+        telemetry_type: TelemetryType = TelemetryType(metric_info.telemetry_type)
+        # TODO: Filter out invalid configs: telemetry type, operand
+        # TODO: Rename exception fields
+        metric_info_list = metric_infos.get(telemetry_type, [])
+        metric_info_list.append(metric_info)
+        metric_infos[telemetry_type] = metric_info_list
+        seen_ids.add(metric_info.id)
+        # Initialize projections from this derived metric info
+        _init_derived_metric_projection(metric_info)
+    _set_quickpulse_derived_metric_infos(metric_infos)
+    # Update new etag
+    _set_quickpulse_etag(etag)
+
+
+# Called by record_span/record_log when processing a span/log_record
+# Derives metrics from projections if applicable to current filters in config
+def _derive_metrics_from_telemetry_data(data: _TelemetryData):
+    metric_infos_dict = _get_quickpulse_derived_metric_infos()
+    metric_infos = []
+    if isinstance(data, _RequestData):
+        metric_infos = metric_infos_dict.get(TelemetryType.REQUEST)
+    elif isinstance(data, _DependencyData):
+        metric_infos = metric_infos_dict.get(TelemetryType.DEPENDENCY)
+    elif isinstance(data, _ExceptionData):
+        metric_infos = metric_infos_dict.get(TelemetryType.EXCEPTION)
+    elif isinstance(data, _TraceData):
+        metric_infos = metric_infos_dict.get(TelemetryType.TRACE)
+    if metric_infos and _check_metric_filters(metric_infos, data):
+        # Since this data matches the filter, create projections used to
+        # generate filtered metrics
+        _create_projections(metric_infos, data)
+    # TODO: Configuration error handling
+
+
+def _check_metric_filters(metric_infos: List[DerivedMetricInfo], data: _TelemetryData) -> bool:
+    match = False
+    for metric_info in metric_infos:
+        # Should only be a single `FilterConjunctionGroupInfo` in `filter_groups`
+        # but we use a logical OR to match if there is more than one
+        for group in metric_info.filter_groups:
+            match = match or _check_filters(group.filters, data)
+    return match
+
+
+def _check_filters(filters: List[FilterInfo], data: _TelemetryData) -> bool:
```

Comment:
nit: You may want to change the type of filters to FilterConjunctionGroupInfo ~ FilterConjunctionGroupInfo is the common denominator between documents/metrics filtering can be helpful for future code reuse.

Reply:
Will keep this for now and refactor once we starts documents. It is not immediately obvious to me the most optimal way to do this.